### ISSUE TITLE
Define where source attestations are stored

### ISSUE THREAD

**Issue by TomHennen** [slsa-framework/slsa#1129]
We should provide guidance on where source attestations are stored.

We may not want to be too prescriptive but should provide an allowance for how a source control system (#1128) should do so.

**Comment by marcelamelara** [comment]
I think this is a general question for SLSA. @paveliak and I were having this exact discussion about the build environment attestations today as well. We also don't want to be too prescriptive, but provide some practical recommendations.

**Comment by TomHennen** [comment]
Perhaps we should require the Source Control System (or Build system in the case of the build environment attestations) to document where they're stored and then provide examples in the sample profiles (#1142)?

For `git` we might suggest "store in the VCS like so", for other systems maybe we suggest a "content addressable storage system by <repository identifier>/<revision id>".

**Comment by TomHennen** [comment]
My inclination is to say something like

"Systems which implement SLSA SHOULD store attestations about attested artifacts adjacent to the artifacts themselves. This aligns the failure domain of attestation distribution with artifact distribution and it helps align read ACLs of attestations with the artifacts. This makes it easier for consumers to use attestations as needed.

In all cases Systems MUST document where consumers can find the relevant attestations for the artifacts produced by the system.

Systems MAY replace more detailed attestations (e.g. provenance) with Verification Summary Attestations (VSAs) as needed for ease-of-use, performance, confidentiality, or any other reason, unless forbidden by the relevant SLSA track."

WDYT @marcelamelara @adityasaky @zachariahcox ?

**Comment by TomHennen** [comment]
@adityasaky I'd especially like any thoughts you have on where these can be stored for git revisions :)

**Comment by zachariahcox** [comment]
This is an interesting question. 
It makes sense to say "if you can read the artifact you must be able to find and read the attestations"

Issuers can come from anywhere though and they could all store things in different places -- I'm not sure how far we can go with saying things should be adjacent? 

Should this just be "organizations must explain where to find 'org-issued provenance' for their artifacts"? 
That could easily be "in the same folder as the artifact" for some orgs, or it could be pgi sigstore

**Comment by TomHennen** [comment]
"Issuers can come from anywhere though"

Oh that's interesting, are you thinking of third parties outside the organization responsible for the attested artifact or are you thinking multiple parties within the organization?

**Comment by adityasaky** [comment]
> Systems which implement SLSA SHOULD store attestations about attested artifacts adjacent to the artifacts themselves. This aligns the failure domain of attestation distribution with artifact distribution and it helps align read ACLs of attestations with the artifacts. This makes it easier for consumers to use attestations as needed.
>
> In all cases Systems MUST document where consumers can find the relevant attestations for the artifacts produced by the system.

I think this is great!

> Systems MAY replace more detailed attestations (e.g. provenance) with Verification Summary Attestations (VSAs) as needed for ease-of-use, performance, confidentiality, or any other reason, unless forbidden by the relevant SLSA track.

I'm less sure about this. Here, we're discussing the production and storing of attestations, and not their use by a verifier. It's one thing to use a VSA in place of a full attestation during verification somewhere (say under resource constraints) but the proposed text reads like the full attestation may not be produced / stored at all. IMO, a VSA should be backed by full attestations, as it is contingent on the verification of some set of facts. While these set of facts don't have to be attestations, there's no reason they can't just be full attestations?

> I'd especially like any thoughts you have on where these can be stored for git revisions :)

We store attestations in gittuf in `refs/gittuf/attestations` and use them during policy verification, primarily for threshold approval of changes. As a datapoint, [gitsign](https://github.com/sigstore/gitsign) also has the ability to generate attestations and store them in custom references `refs/attestations/commits` and `refs/attestations/trees`.

Storing them in the repository using custom references makes sense for gittuf because the idea is to have a repository be "self contained" with respect to its policies and activity information, which allows for independent verification of the repository's security controls. It meets the "store alongside the artifacts" recommendation, but it can lead to a repository's size ballooning depending on how much data any single attestation contains (I think @zachariahcox mentioned this on a call too).

More generally, I think the text saying that a) the attestations must be stored next to the artifacts (thus, making it part of the SCS), and b) the system must document where to get the attestations is sufficient. This may be using Git revisions or this may be using something like GitHub's artifact attestations feature, or even both depending on the expected verification contexts.

> > Issuers can come from anywhere though and they could all store things in different places -- I'm not sure how far we can go with saying things should be adjacent?
>
> Oh that's interesting, are you thinking of third parties outside the organization responsible for the attested artifact or are you thinking multiple parties within the organization?

I'm also curious to hear from @zachariahcox on this. :)

FWIW, I view this as a producer vs consumer question. If as the producer of some software, I have third parties attesting to my software, I think the third parties are still part of my overall SCS and I can likely find a way for these attestations to be collated and provided along with my repository artifacts? Does scorecard, for example, already fit this definition? On the other hand, as a consumer of some upstream repository, I may use the attestations provided alongside the repository _and_ attestations from some third party, say some sort of OSS monitoring or reputation scoring service. In this case, I wouldn't expect the third party attestations to be provided alongside the repository artifacts because it's not part of the producer's SCS. In my mind, this track is focused on software producers, so I think we may be able to make this recommendation?

**Comment by zachariahcox** [comment]
Howdy friends! Catching up from two weeks of conferences. 

Generally this problem feels like it should not require we define "how to publish correctness policies." Typically I need to use this when I want to know if an artifact is correct -- i'll go download all the attestations from all the places. 
There could be malicious or incorrect attestations on the internet, and I need to know NOT to depend on those somehow. That's related to knowing "what does correct look like for this artifact" and "where do the correctness evaluators publish their attestations?"

**Comment by TomHennen** [comment]
Chatting about this at the SLSA spec meeting today and it was broadly agreed that this shouldn't be a blocker to release of the source track.  I'll leave it open as we'll still want something for it, but I will remove it from the project board if it's there.

**Comment by zachariahcox** [comment]
> but I will remove it from the project board if it's there.

I vote to just close it -- we can always reopen it later if we decide we need it.
